homeassistant/components/glances/coordinator.py

```diff
@@ -42,4 +44,31 @@ async def _async_update_data(self) -> dict[str, Any]:
             raise ConfigEntryAuthFailed from err
         except exceptions.GlancesApiError as err:
             raise UpdateFailed from err
+        # Update computed values
+        if data:
```

Comment:
How about just checking here if it should refresh. Something like:
```python
if data and self._last_refresh_uptime < dt_util.utcnow() - UPTIME_REFRESH_RATE
```

Reply:
The test is a bit more complex, because we don't just wait for a certain time to refresh. We need to check if the uptime itself has changed significantly, and update it only of it has moved more than a certain amount.
Hence the comparison to the previous uptime value, not to the previous refresh time.

What we want to avoid is a succession of changes to uptime like this (what i called flapping) - those are actual values for a server that was not restarted:
 uptime=2024-03-10 21:**18**:28.**588024**+00:00
 uptime=2024-03-10 21:**19**:28.**397835**+00:00
 uptime=2024-03-10 21:**20**:28.**381627**+00:00
 uptime=2024-03-10 21:**18**:28.**408223**+00:00
 uptime=2024-03-10 21:**19**:28.**393393**+00:00
 uptime=2024-03-10 21:**20**:28.**391400**+00:00
 uptime=2024-03-10 21:**18**:28.**410554**+00:00
 uptime=2024-03-10 21:**19**:28.**382378**+00:00

Reply:
So actually if we don't refresh in 10 minutes the next update might be still setting a datetime which is earlier than the previous one? That seems highly strange and I wonder why we should even have it all then?

So it's actually not uptime but instead some start time or what is actually this value representing as it can change as it does and is not a starting time?

Reply:
Ref: https://glances.readthedocs.io/en/latest/api.html#get-uptime
I would expect the api to come back with something like `12 days, 3:42:29`?

Reply:
Yes the API returns a timespan since last reboot of the remote server like the string you give.
However it does not provide the time at which this timespan has been calculated on the remote server, so when we convert it to an absolute datetime in HA, we use the current time `utcnow()`: `uptime=utcnow()-timespan`
Depending on how much time it took for the timespan to reach the HA server, the resulting absolute datetime varies slightly at each update, even if the remote server has not been rebooted, and it is this slight variation that we need to eliminate.
In the example I gave, the server is far away with a low polling frequency (internal to Glances) so the variations can be several minutes. Typically it is often a few seconds.

Reply:
Use https://glances.readthedocs.io/en/latest/api.html#get-now
Then use that instead of `utcnow()` to get a proper start time. 

Also it would be enough to calculate this once until the server has restarted. Possible to get in some easy manner?

Reply:
The sensor `now` is not provided by the current implementation of glances-api, I would need to ask the library to add it.
https://github.com/home-assistant-ecosystem/python-glances-api/blob/a73164393868d6c76421208953e771b9297209ed/glances_api/__init__.py#L100

This would improve greatly the situation, there would still be a bit of variation, I would need to test.

As it seems this part of the code is the blocking point, I can suggest an approach in 2 steps:
- I can remove the code that checks the new uptime versus the previous uptime value. Set uptime seconds to 0 in the datetime, to keep only date/hours/minutes. This should remove most flapping issues (except when a server starts very close to the minute mark)
- In a new PR for a following release, swap the `utcnow()` call by the `now` sensor once it is provided by the API (I can create the follow up PRs in both the API and HA)

Let me know if this would be acceptable?

Thanks!

Reply:
Also I don't think there is any way to know that the server restarted, to compute the uptime only once. I haven't found any data in the API which would change at each restart.

Reply:
I would suggest we draft this PR and get that implemented into the library and then we can bump that here. 
I acknowledge it delays this PR but I wouldn't like to implement it in its current way anyhow. Thanks

Reply:
Ok I'll do that, thanks

Reply:
Issue:
https://github.com/home-assistant-ecosystem/python-glances-api/issues/33

Reply:
As mentioned below, using `now` doesn't solve the question of small variations, and also introduces other difficulties with timezone support. 
I believe using `utcnow()`, like other integrations in HA, is a safer and simpler approach.

Reply:
As discussed, I have changed the logic to only update uptime on startup or when remote server restarts